IRGen: Invoke objc_opt_self directly when available.

jckarter opened the pull request:
We don't need swift_getInitializedObjCClass on new enough Apple OSes because
the ObjC runtime provides an equivalent call for us.

jckarter committed:
IRGen: Invoke objc_opt_self directly when available.

We don't need swift_getInitializedObjCClass on new enough Apple OSes because
the ObjC runtime provides an equivalent call for us.

jckarter commented:
@swift-ci Please test

compnerd commented on lib/IRGen/GenHeap.cpp:
```diff
@@ -1244,6 +1244,32 @@ llvm::Constant *IRGenModule::getFixLifetimeFn() {
+  // If ObjC interop is disabled, we don't need to do fixed class
+  // initialization.
+  llvm::Constant *fn;
+  if (!ObjCInterop) {
```
I think that this could be hoisted further:

```
  if (FixedClassInitializationFn)
    return *FixedClassInitializationFn;

  if (!ObjcInterop)
    return nullptr;

  if (getAvailabilityContext().isContainedIn(Context.getSwift51Availability()))
    FixedClassInitializationFn = getObjCOptSelFn();
  else
    FixedClassInitializationFn = getGetInitializedObjCClassFn();

  return FixedClassInitializationFn;
```

jckarter replied:
Because FixedClassInitializationFn is  Optional, and is a nonlocal instance variable, I think it’s preferable to keep the new value in a nonoptional local variable and assign to global state with a weaker type once.

jckarter commented:
@swift-ci Please test

jckarter committed:
Remove explicit target(s) from reflect_empty_struct tests

Swift CI tests backward deployment on different host OSes, so this should be covered in
testing. The explicit target doesn't work well with `executable_test` since it leads to
binaries with too-new minimum deployment targets getting executed on older host OSes.

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test

jckarter commented:
@swift-ci Please test
